Fix an error when specify `--define=foo`

ydah opened the pull request:
```
❯ exe/lrama --define=foo sample/calc.y
wrong array length at 0 (expected 2, was 1)
```

This is occurring and has been addressed below:
https://github.com/ruby/lrama/blob/7fa74d84a68b3632950b969d29ccf012a0ea4a42/lib/lrama/grammar.rb#L60

yui-knk commented:
I'm not sure why steep didn't warn however please update https://github.com/ruby/lrama/blob/7fa74d84a68b3632950b969d29ccf012a0ea4a42/sig/lrama/options.rbs#L4 too.

yui-knk commented:
[Note] It seems steep doesn't warn on array assignment to Hash type instance variable:

```ruby
# code
@define = []

# rbs
attr_accessor define: Hash[String, String]
```

```shell
$ bundle exec steep check
# Type checking files:

......................................................................................................................

No type error detected. 🫖
```

but it warns on hash assignment to Array type instance variable:

```ruby
# code
@define = {}

# rbs
attr_accessor define: Array[String]
```

```shell
$ bundle exec steep check
# Type checking files:

.....................................................................F................................................

lib/lrama/options.rb:15:16: [warning] Empty hash doesn't have type annotation
│ Diagnostic ID: Ruby::UnannotatedEmptyCollection
│
└       @define = {}
                  ~~

Detected 1 problem from 1 file
```

junk0612 commented:
This change alone is probably not sufficient, and changes are needed, for example, in the location of the assignment at below. https://github.com/ruby/lrama/blob/7fa74d84a68b3632950b969d29ccf012a0ea4a42/parser.y#L28
By the way, I think Hash is appropriate for define because it is written in the grammar file as `%define lr.type ielr` or from the command line as `lrama -D lr.type=ielr`, but why change it to Array?

ydah commented:
> This change alone is probably not sufficient, and changes are needed, for example, in the location of the assignment at below.

Changes there will be made separately. Certainly as it is now, if a value is missing, it will result in an error.

> By the way, I think Hash is appropriate for define because it is written in the grammar file as `%define lr.type ielr` or from the command line as `lrama -D lr.type=ielr`, but why change it to Array?

I didn't mean to change the `@define` to Array in the first place. Because, below, I am converting from Array to Hash.
https://github.com/ruby/lrama/blob/7fa74d84a68b3632950b969d29ccf012a0ea4a42/lib/lrama/grammar.rb#L60

This PR simply changes the default value from Hash to Array.
Because the following Option is received in Array.
https://github.com/ruby/lrama/blob/968725eec15d1c0967d6bc75490febf78d66fcae/lib/lrama/option_parser.rb#L63

If I understand something incorrectly, please point it out to me.

ydah commented:
I changed it back to Hash to avoid the error https://github.com/ruby/lrama/pull/527/commits/569f568853e051db866f929ccf83fa8b71d7b290
And moved to rbs-inline style type definition https://github.com/ruby/lrama/pull/527/commits/4bd35c084b9aa1324ff50097349e77473195d1ac

yui-knk commented on lib/lrama/options.rb:
```diff
@@ -1,15 +1,27 @@
+    attr_accessor :skeleton #: String
+    attr_accessor :header #: bool
+    attr_accessor :header_file #: String
+    attr_accessor :report_file #: String
```
It's initialized with `nil` then `String?`.

ydah replied:
Fixed: https://github.com/ruby/lrama/pull/527/commits/7e064dea26300767aa3d30e7fb49a72985c8a55b

yui-knk commented on lib/lrama/options.rb:
```diff
@@ -1,15 +1,27 @@
+    attr_accessor :report_file #: String
+    attr_accessor :outfile #: String
+    attr_accessor :error_recovery #: bool
+    attr_accessor :grammar_file #: String
```
It's initialized with `nil` then `String?`.

ydah replied:
Fixed: https://github.com/ruby/lrama/pull/527/commits/7e064dea26300767aa3d30e7fb49a72985c8a55b

yui-knk commented on spec/lrama/option_parser_spec.rb:
```diff
@@ -343,4 +343,24 @@
     end
   end
+
+  describe "@define" do
```
Could you add a test case for the issue which this PR will resolve, in short test case for `--define=foo`?

ydah replied:
Added: https://github.com/ruby/lrama/pull/527/commits/5dfb6a3f77e7877a3dc962bcb05efd6c50889b11

junk0612 commented:
> I didn't mean to change the @define to Array in the first place. Because, below, I am converting from Array to Hash. It looks like the intended code.

I see, indeed. My understanding was wrong.

ydah commented on lib/lrama/options.rb:
```diff
@@ -1,15 +1,27 @@
-                  :diagram, :diagram_file, :profile_opts
+    attr_accessor :skeleton #: String
+    attr_accessor :header #: bool
+    attr_accessor :header_file #: String
```
Fixed: https://github.com/ruby/lrama/pull/527/commits/7e064dea26300767aa3d30e7fb49a72985c8a55b
